src/rtutils.c

```diff
@@ -297,6 +297,16 @@ JL_DLLEXPORT void jl_eh_restore_state(jl_handler_t *eh)
     }
 }
 
+JL_DLLEXPORT void jl_eh_restore_state_noexcept(jl_handler_t *eh)
+{
+    jl_task_t *ct = jl_current_task;
+    ct->eh = eh->prev;
+    ct->gcstack = eh->gcstack;
+    ct->world_age = eh->world_age;
+    ct->ptls->defer_signal = eh->defer_signal;
+    return;
+}
+
 JL_DLLEXPORT void jl_pop_handler(int n)
```

Comment:
The PR LGTM. Just a note though that codegen always calls this function exclusively, even though it knows `n` is constant and probably also therefore usually knows `eh`. Which means it may be able to optimize that performance by calling `jl_eh_restore_state_noexcept` directly. But in the meantime, jl_pop_handler should call this new function too and be marked JL_NOTSAFEPOINT. Similarly, in interpreter.c, you can split the call to `jl_eh_restore_state` into the `continue_at` and `catch` cases there for consistency with this.